apps/studio/src/server/modules/resource/resource.service.ts

```diff
@@ -231,3 +233,147 @@ export const moveResource = async (
+
+    ancestors.push(parent)
+    currentResource = parent
+  }
+
+  // Step 3: Get the immediate children of the resource
+  const children = await db
+    .selectFrom("Resource")
+    .where("Resource.siteId", "=", siteId)
+    .where("Resource.parentId", "=", String(resourceId))
+    .select(defaultResourceSelect)
+    .execute()
+
+  // Step 4: Get the immediate siblings of the resource
+  const siblingsQuery = db
+    .selectFrom("Resource")
+    .where("Resource.siteId", "=", siteId)
+    .where("Resource.id", "!=", String(resourceId))
+    .select(defaultResourceSelect)
+
+  if (resource.parentId === null) {
+    siblingsQuery.where("Resource.parentId", "is", null)
+  } else {
+    siblingsQuery.where("Resource.parentId", "=", resource.parentId)
+  }
+
+  const siblings = await siblingsQuery.execute()
+
+  // Step 5: Get the immediate children of the siblings
+  const siblingIds = siblings.map((s) => s.id)
+  let siblingChildrenMap = {}
+
+  if (siblingIds.length > 0) {
+    const siblingChildren = await db
+      .selectFrom("Resource")
+      .where("Resource.siteId", "=", siteId)
+      .where("Resource.parentId", "in", siblingIds)
+      .select(defaultResourceSelect)
+      .execute()
+
+    siblingChildrenMap = siblingChildren.reduce<
+      Record<string, typeof siblingChildren>
+    >((acc, child) => {
+      // NOTE: This case will never happen, because we have already filtered
+      // specifically for children of the above siblings
+      if (child.parentId === null) {
+        return acc
+      }
+
+      acc[child.parentId] = [...(acc[child.parentId] ?? []), child]
+
+      return acc
+    }, {})
+  }
+
+  // Step 6: Construct the localised sitemap object
+  const rootResource = await db
+    .selectFrom("Resource")
+    .where("Resource.siteId", "=", siteId)
+    .where("Resource.parentId", "is", null)
+    .where("Resource.permalink", "=", "")
+    .select(defaultResourceSelect)
+    .executeTakeFirst()
+
+  if (rootResource === undefined) {
+    // This case will never happen, because we have guaranteed that there is
+    // always the root resource
+    throw new Error("Root item not found")
+  }
+
+  const allResources = [...ancestors, ...siblings, ...children]
+
+  return getSitemapTree(rootResource, allResources)
+}
+
+export const getResourceFullPermalink = async (
+  siteId: number,
+  resourceId: number,
+) => {
+  // Step 1: Get the actual resource
+  const resource = await getById(db, { resourceId, siteId })
+    .select(defaultResourceSelect)
+    .executeTakeFirstOrThrow()
+
+  // Step 2: Get all the ancestors of the resource
+  const ancestors = [resource]
+  let currentResource = resource
+
+  while (currentResource.parentId !== null) {
+    const parent = await getById(db, {
+      resourceId: Number(currentResource.parentId),
+      siteId,
+    })
+      .select(defaultResourceSelect)
+      .executeTakeFirstOrThrow()
+
+    ancestors.push(parent)
+    currentResource = parent
```

Comment:
same comment as earlier, might be useful to do a recursive cte here

Reply:
Changed in bd051a9c8f9d1a82abde4ceb4c108498fb00a10a.